include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1160,34 +1171,257 @@ __parallel_set_scan(sycl::queue& __q, _Range1&& __rng1, _Range2&& __rng2, _Range
+        __q, __tmp_rng3.all_view(), __tmp_rng4.all_view(), std::forward<_Range3>(__result), __comp)
+        .wait();
+    return __n_diff_1 + __n_diff_2;
+}
+
+template <typename _CustomName, typename _Range1, typename _Range2, typename _Range3, typename _Compare,
+          typename _UseReduceThenScan>
+std::size_t
+__set_write_a_only_op(sycl::queue& __q, _Range1&& __rng1, _Range2&& __rng2, _Range3&& __result, _Compare __comp,
+                      oneapi::dpl::unseq_backend::_IntersectionTag, _UseReduceThenScan)
+{
+    if constexpr (_UseReduceThenScan::value)
+        return __parallel_set_reduce_then_scan_set_a_write<_CustomName>(
+                   __q, std::forward<_Range1>(__rng1), std::forward<_Range2>(__rng2), std::forward<_Range3>(__result),
+                   __comp, oneapi::dpl::unseq_backend::_IntersectionTag{})
+            .get();
+    else
+        return __parallel_set_scan<_CustomName>(__q, std::forward<_Range1>(__rng1), std::forward<_Range2>(__rng2),
+                                                std::forward<_Range3>(__result), __comp,
+                                                oneapi::dpl::unseq_backend::_IntersectionTag{})
+            .get();
+}
+
+template <typename _CustomName, typename _Range1, typename _Range2, typename _Range3, typename _Compare,
+          typename _UseReduceThenScan>
+std::size_t
+__set_write_a_only_op(sycl::queue& __q, _Range1&& __rng1, _Range2&& __rng2, _Range3&& __result, _Compare __comp,
+                      oneapi::dpl::unseq_backend::_DifferenceTag, _UseReduceThenScan)
+{
+    if constexpr (_UseReduceThenScan::value)
+        return __parallel_set_reduce_then_scan_set_a_write<_CustomName>(
+                   __q, std::forward<_Range1>(__rng1), std::forward<_Range2>(__rng2), std::forward<_Range3>(__result),
+                   __comp, oneapi::dpl::unseq_backend::_DifferenceTag{})
+            .get();
+    else
+        return __parallel_set_scan<_CustomName>(__q, std::forward<_Range1>(__rng1), std::forward<_Range2>(__rng2),
+                                                std::forward<_Range3>(__result), __comp,
+                                                oneapi::dpl::unseq_backend::_DifferenceTag{})
+            .get();
 }
 
-template <typename _ExecutionPolicy>
+template <typename _CustomName>
+struct reduce_then_scan_wrapper;
+
+template <typename _CustomName>
+struct scan_then_propagate_wrapper;
+
+template <typename _CustomName>
+struct set_a_write_wrapper;
+
+template <typename _SetTag, typename _Rng1, typename _Rng2>
 bool
-__can_set_op_write_from_set_b(oneapi::dpl::__internal::__device_backend_tag, _ExecutionPolicy&& __exec)
+__use_write_a_alg(_SetTag, _Rng1&& __rng1, _Rng2&&)
 {
-    return oneapi::dpl::__par_backend_hetero::__is_gpu_with_reduce_then_scan_sg_sz(__exec.queue());
+    using __value_t = oneapi::dpl::__internal::__value_t<_Rng1>;
+    return __rng1.size() < 32 * 1024 * sizeof(__value_t);
```

Comment:
`32 * 1024` - what this mean and why you are using this magic value in this and in some another places?
Probably some comments needed and may be make sense to declare this value as some `constexp` value with required additional comments...

Reply:
refactored to combine and added comments.